File: string-algorithms/edit_distance.go
/**
 * Edit Distance Algorithms - Go Implementation
 * =============================================
 *
 * String similarity and edit distance metrics.
 *
 * Run: go run edit_distance.go
 */

package main

import (
	"fmt"
	"math"
	"strings"
)

// ============================================================================
// Edit Distance Algorithms
// ============================================================================

// HammingDistance computes the Hamming distance between two strings
// Time: O(n), Space: O(1)
func HammingDistance(str1, str2 string) (int, error) {
	if len(str1) != len(str2) {
		return 0, fmt.Errorf("hamming distance requires equal length strings")
	}

	distance := 0
	for i := 0; i < len(str1); i++ {
		if str1[i] != str2[i] {
			distance++
		}
	}

	return distance, nil
}

// LevenshteinDistance computes the Levenshtein distance
// Time: O(nm), Space: O(nm)
func LevenshteinDistance(str1, str2 string, visualize bool) int {
	m, n := len(str1), len(str2)

	// Create DP table
	dp := make([][]int, m+1)
	for i := range dp {
		dp[i] = make([]int, n+1)
	}

	// Initialize base cases
	for i := 0; i <= m; i++ {
		dp[i][0] = i
	}
	for j := 0; j <= n; j++ {
		dp[0][j] = j
	}

	// Fill DP table
	for i := 1; i <= m; i++ {
		for j := 1; j <= n; j++ {
			cost := 0
			if str1[i-1] != str2[j-1] {
				cost = 1
			}

			dp[i][j] = min3(
				dp[i-1][j]+1,      // deletion
				dp[i][j-1]+1,      // insertion
				dp[i-1][j-1]+cost, // substitution
			)
		}
	}

	if visualize {
		visualizeDPTable(str1, str2, dp)
	}

	return dp[m][n]
}

// LevenshteinOptimized computes Levenshtein distance with O(min(n,m)) space
func LevenshteinOptimized(str1, str2 string) int {
	// Ensure str1 is shorter
	if len(str1) > len(str2) {
		str1, str2 = str2, str1
	}

	m, n := len(str1), len(str2)

	// Use two rows
	prevRow := make([]int, n+1)
	currRow := make([]int, n+1)

	for j := 0; j <= n; j++ {
		prevRow[j] = j
	}

	for i := 1; i <= m; i++ {
		currRow[0] = i

		for j := 1; j <= n; j++ {
			cost := 0
			if str1[i-1] != str2[j-1] {
				cost = 1
			}

			currRow[j] = min3(
				prevRow[j]+1,      // deletion
				currRow[j-1]+1,    // insertion
				prevRow[j-1]+cost, // substitution
			)
		}

		prevRow, currRow = currRow, prevRow
	}

	return prevRow[n]
}

// DamerauLevenshteinDistance computes Damerau-Levenshtein distance
// Time: O(nm), Space: O(nm)
func DamerauLevenshteinDistance(str1, str2 string) int {
	m, n := len(str1), len(str2)
	maxDist := m + n

	// Use map for sparse matrix
	H := make(map[[2]int]int)

	// Initialize
	H[[2]int{-1, -1}] = maxDist
	for i := 0; i <= m; i++ {
		H[[2]int{i, -1}] = maxDist
		H[[2]int{i, 0}] = i
	}
	for j := 0; j <= n; j++ {
		H[[2]int{-1, j}] = maxDist
		H[[2]int{0, j}] = j
	}

	for i := 1; i <= m; i++ {
		DB := 0

		for j := 1; j <= n; j++ {
			k := DB
			l := 0
			if str1[i-1] != str2[j-1] {
				l = 1
			} else {
				DB = j
			}

			H[[2]int{i, j}] = min4(
				H[[2]int{i-1, j}]+1,                                    // deletion
				H[[2]int{i, j-1}]+1,                                    // insertion
				H[[2]int{i-1, j-1}]+l,                                  // substitution
				H[[2]int{k-1, DB-1}]+(i-k-1)+1+(j-DB-1), // transposition
			)
		}
	}

	return H[[2]int{m, n}]
}

// LCSLength computes the length of Longest Common Subsequence
// Time: O(nm), Space: O(nm)
func LCSLength(str1, str2 string) int {
	m, n := len(str1), len(str2)

	dp := make([][]int, m+1)
	for i := range dp {
		dp[i] = make([]int, n+1)
	}

	for i := 1; i <= m; i++ {
		for j := 1; j <= n; j++ {
			if str1[i-1] == str2[j-1] {
				dp[i][j] = dp[i-1][j-1] + 1
			} else {
				dp[i][j] = max(dp[i-1][j], dp[i][j-1])
			}
		}
	}

	return dp[m][n]
}

// LCSString returns the actual Longest Common Subsequence
func LCSString(str1, str2 string) string {
	m, n := len(str1), len(str2)

	dp := make([][]int, m+1)
	for i := range dp {
		dp[i] = make([]int, n+1)
	}

	for i := 1; i <= m; i++ {
		for j := 1; j <= n; j++ {
			if str1[i-1] == str2[j-1] {
				dp[i][j] = dp[i-1][j-1] + 1
			} else {
				dp[i][j] = max(dp[i-1][j], dp[i][j-1])
			}
		}
	}

	// Backtrack to find LCS
	var lcs strings.Builder
	i, j := m, n

	for i > 0 && j > 0 {
		if str1[i-1] == str2[j-1] {
			lcs.WriteByte(str1[i-1])
			i--
			j--
		} else if dp[i-1][j] > dp[i][j-1] {
			i--
		} else {
			j--
		}
	}

	// Reverse the string
	result := lcs.String()
	return reverse(result)
}

// EditOperation represents an edit operation
type EditOperation struct {
	Op   string
	Pos  int
	Char byte
}

// EditSequence returns the sequence of edit operations
func EditSequence(str1, str2 string) []EditOperation {
	m, n := len(str1), len(str2)

	dp := make([][]int, m+1)
	ops := make([][]*EditOperation, m+1)
	for i := range dp {
		dp[i] = make([]int, n+1)
		ops[i] = make([]*EditOperation, n+1)
	}

	// Initialize
	for i := 0; i <= m; i++ {
		dp[i][0] = i
		if i > 0 {
			ops[i][0] = &EditOperation{"delete", i - 1, str1[i-1]}
		}
	}

	for j := 0; j <= n; j++ {
		dp[0][j] = j
		if j > 0 {
			ops[0][j] = &EditOperation{"insert", j - 1, str2[j-1]}
		}
	}

	// Fill tables
	for i := 1; i <= m; i++ {
		for j := 1; j <= n; j++ {
			if str1[i-1] == str2[j-1] {
				dp[i][j] = dp[i-1][j-1]
				ops[i][j] = &EditOperation{"match", i - 1, str1[i-1]}
			} else {
				costs := []struct {
					cost int
					op   EditOperation
				}{
					{dp[i-1][j] + 1, EditOperation{"delete", i - 1, str1[i-1]}},
					{dp[i][j-1] + 1, EditOperation{"insert", j - 1, str2[j-1]}},
					{dp[i-1][j-1] + 1, EditOperation{"substitute", i - 1, str2[j-1]}},
				}

				minCost := costs[0]
				for _, c := range costs[1:] {
					if c.cost < minCost.cost {
						minCost = c
					}
				}

				dp[i][j] = minCost.cost
				ops[i][j] = &minCost.op
			}
		}
	}

	// Backtrack
	sequence := []EditOperation{}
	i, j := m, n

	for i > 0 || j > 0 {
		if ops[i][j] != nil {
			sequence = append([]EditOperation{*ops[i][j]}, sequence...)

			switch ops[i][j].Op {
			case "match", "substitute":
				i--
				j--
			case "delete":
				i--
			case "insert":
				j--
			}
		} else {
			break
		}
	}

	return sequence
}

// ============================================================================
// Similarity Metrics
// ============================================================================

// NormalizedLevenshtein computes normalized Levenshtein distance (0 to 1)
func NormalizedLevenshtein(str1, str2 string) float64 {
	if len(str1) == 0 && len(str2) == 0 {
		return 1.0
	}

	distance := LevenshteinDistance(str1, str2, false)
	maxLen := max(len(str1), len(str2))

	return 1.0 - float64(distance)/float64(maxLen)
}

// SimilarityRatio computes similarity ratio based on LCS
func SimilarityRatio(str1, str2 string) float64 {
	if len(str1) == 0 && len(str2) == 0 {
		return 1.0
	}

	lcsLen := LCSLength(str1, str2)
	totalLen := len(str1) + len(str2)

	if totalLen == 0 {
		return 1.0
	}

	return 2.0 * float64(lcsLen) / float64(totalLen)
}

// JaroDistance computes Jaro distance
func JaroDistance(str1, str2 string) float64 {
	if str1 == str2 {
		return 1.0
	}

	len1, len2 := len(str1), len(str2)

	if len1 == 0 || len2 == 0 {
		return 0.0
	}

	// Maximum allowed distance
	matchDistance := max(len1, len2)/2 - 1
	if matchDistance < 1 {
		matchDistance = 1
	}

	str1Matches := make([]bool, len1)
	str2Matches := make([]bool, len2)

	matches := 0
	transpositions := 0

	// Find matches
	for i := 0; i < len1; i++ {
		start := max(0, i-matchDistance)
		end := min(i+matchDistance+1, len2)

		for j := start; j < end; j++ {
			if str2Matches[j] || str1[i] != str2[j] {
				continue
			}
			str1Matches[i] = true
			str2Matches[j] = true
			matches++
			break
		}
	}

	if matches == 0 {
		return 0.0
	}

	// Find transpositions
	k := 0
	for i := 0; i < len1; i++ {
		if !str1Matches[i] {
			continue
		}
		for !str2Matches[k] {
			k++
		}
		if str1[i] != str2[k] {
			transpositions++
		}
		k++
	}

	return (float64(matches)/float64(len1) +
		float64(matches)/float64(len2) +
		float64(matches-transpositions/2)/float64(matches)) / 3.0
}

// JaroWinklerDistance computes Jaro-Winkler distance
func JaroWinklerDistance(str1, str2 string, prefixScale float64) float64 {
	jaro := JaroDistance(str1, str2)

	// Find common prefix length (max 4)
	prefix := 0
	minLen := min(len(str1), len(str2))
	if minLen > 4 {
		minLen = 4
	}

	for i := 0; i < minLen; i++ {
		if str1[i] == str2[i] {
			prefix++
		} else {
			break
		}
	}

	return jaro + float64(prefix)*prefixScale*(1-jaro)
}

// ============================================================================
// Helper Functions
// ============================================================================

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min3(a, b, c int) int {
	return min(min(a, b), c)
}

func min4(a, b, c, d int) int {
	return min(min3(a, b, c), d)
}

func reverse(s string) string {
	runes := []rune(s)
	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	return string(runes)
}

func visualizeDPTable(str1, str2 string, dp [][]int) {
	fmt.Println("\nDP Table:")
	fmt.Println(strings.Repeat("=", 70))

	// Header
	fmt.Print("       ")
	for _, c := range str2 {
		fmt.Printf("%4c", c)
	}
	fmt.Println()

	// Rows
	for i, row := range dp {
		if i == 0 {
			fmt.Print("  ")
		} else {
			fmt.Printf("%2c", str1[i-1])
		}

		for _, val := range row {
			fmt.Printf("%4d", val)
		}
		fmt.Println()
	}
}

// ============================================================================
// Main Function - Examples
// ============================================================================

func main() {
	fmt.Println(strings.Repeat("=", 70))
	fmt.Println("EDIT DISTANCE ALGORITHMS - GO")
	fmt.Println(strings.Repeat("=", 70))

	// Example 1: Hamming distance
	fmt.Println("\n" + strings.Repeat("=", 70))
	fmt.Println("EXAMPLE 1: Hamming Distance")
	fmt.Println(strings.Repeat("=", 70))

	str1, str2 := "karolin", "kathrin"
	dist, err := HammingDistance(str1, str2)
	if err != nil {
		fmt.Println("Error:", err)
	} else {
		fmt.Printf("String 1: '%s'\n", str1)
		fmt.Printf("String 2: '%s'\n", str2)
		fmt.Printf("Hamming distance: %d\n", dist)
	}

	// Example 2: Levenshtein distance
	fmt.Println("\n" + strings.Repeat("=", 70))
	fmt.Println("EXAMPLE 2: Levenshtein Distance")
	fmt.Println(strings.Repeat("=", 70))

	str3, str4 := "kitten", "sitting"
	dist2 := LevenshteinDistance(str3, str4, true)

	fmt.Printf("\nString 1: '%s'\n", str3)
	fmt.Printf("String 2: '%s'\n", str4)
	fmt.Printf("Levenshtein distance: %d\n", dist2)

	// Example 3: Damerau-Levenshtein
	fmt.Println("\n" + strings.Repeat("=", 70))
	fmt.Println("EXAMPLE 3: Damerau-Levenshtein Distance")
	fmt.Println(strings.Repeat("=", 70))

	testCases := [][2]string{
		{"kitten", "sitting"},
		{"teh", "the"},
		{"abcd", "acbd"},
	}

	for _, tc := range testCases {
		s1, s2 := tc[0], tc[1]
		levDist := LevenshteinDistance(s1, s2, false)
		damLevDist := DamerauLevenshteinDistance(s1, s2)

		fmt.Printf("\nString 1: '%s'\n", s1)
		fmt.Printf("String 2: '%s'\n", s2)
		fmt.Printf("  Levenshtein:         %d\n", levDist)
		fmt.Printf("  Damerau-Levenshtein: %d\n", damLevDist)
	}

	// Example 4: LCS
	fmt.Println("\n" + strings.Repeat("=", 70))
	fmt.Println("EXAMPLE 4: Longest Common Subsequence")
	fmt.Println(strings.Repeat("=", 70))

	str5, str6 := "ABCDGH", "AEDFHR"
	lcs := LCSString(str5, str6)

	fmt.Printf("String 1: '%s'\n", str5)
	fmt.Printf("String 2: '%s'\n", str6)
	fmt.Printf("LCS: '%s' (length: %d)\n", lcs, len(lcs))

	// Example 5: Similarity metrics
	fmt.Println("\n" + strings.Repeat("=", 70))
	fmt.Println("EXAMPLE 5: Similarity Metrics")
	fmt.Println(strings.Repeat("=", 70))

	testPairs := [][2]string{
		{"kitten", "sitting"},
		{"saturday", "sunday"},
		{"martha", "marhta"},
		{"dixon", "dicksonx"},
	}

	for _, pair := range testPairs {
		s1, s2 := pair[0], pair[1]
		normLev := NormalizedLevenshtein(s1, s2)
		simRatio := SimilarityRatio(s1, s2)
		jaro := JaroDistance(s1, s2)
		jaroWinkler := JaroWinklerDistance(s1, s2, 0.1)

		fmt.Printf("\n'%s' vs '%s':\n", s1, s2)
		fmt.Printf("  Normalized Levenshtein: %.4f\n", normLev)
		fmt.Printf("  Similarity Ratio:       %.4f\n", simRatio)
		fmt.Printf("  Jaro:                   %.4f\n", jaro)
		fmt.Printf("  Jaro-Winkler:           %.4f\n", jaroWinkler)
	}

	fmt.Println("\n" + strings.Repeat("=", 70))
}

File: string-algorithms/edit_distance.rs
/**
 * Edit Distance Algorithms - Rust Implementation
 * ===============================================
 *
 * String similarity and edit distance metrics.
 *
 * Compile: rustc -O edit_distance.rs
 * Run: ./edit_distance
 */

use std::cmp;
use std::collections::HashMap;

// ============================================================================
// Edit Distance Algorithms
// ============================================================================

/// Hamming Distance - Number of positions at which symbols differ
/// Time: O(n), Space: O(1)
pub fn hamming_distance(str1: &str, str2: &str) -> Result<usize, &'static str> {
    if str1.len() != str2.len() {
        return Err("Hamming distance requires equal length strings");
    }

    let distance = str1
        .chars()
        .zip(str2.chars())
        .filter(|(c1, c2)| c1 != c2)
        .count();

    Ok(distance)
}

/// Levenshtein Distance - Minimum edits (insert, delete, substitute)
/// Time: O(nm), Space: O(nm)
pub fn levenshtein_distance(str1: &str, str2: &str, visualize: bool) -> usize {
    let m = str1.len();
    let n = str2.len();

    // Create DP table
    let mut dp = vec![vec![0; n + 1]; m + 1];

    // Initialize base cases
    for i in 0..=m {
        dp[i][0] = i;
    }
    for j in 0..=n {
        dp[0][j] = j;
    }

    // Convert to char vectors for indexing
    let chars1: Vec<char> = str1.chars().collect();
    let chars2: Vec<char> = str2.chars().collect();

    // Fill DP table
    for i in 1..=m {
        for j in 1..=n {
            let cost = if chars1[i - 1] == chars2[j - 1] { 0 } else { 1 };

            dp[i][j] = cmp::min(
                cmp::min(
                    dp[i - 1][j] + 1,      // deletion
                    dp[i][j - 1] + 1,      // insertion
                ),
                dp[i - 1][j - 1] + cost,   // substitution
            );
        }
    }

    if visualize {
        visualize_dp_table(str1, str2, &dp);
    }

    dp[m][n]
}

/// Space-optimized Levenshtein distance
/// Time: O(nm), Space: O(min(n,m))
pub fn levenshtein_optimized(str1: &str, str2: &str) -> usize {
    // Ensure str1 is shorter
    let (str1, str2) = if str1.len() > str2.len() {
        (str2, str1)
    } else {
        (str1, str2)
    };

    let m = str1.len();
    let n = str2.len();

    let chars1: Vec<char> = str1.chars().collect();
    let chars2: Vec<char> = str2.chars().collect();

    // Use two rows
    let mut prev_row: Vec<usize> = (0..=n).collect();
    let mut curr_row = vec![0; n + 1];

    for i in 1..=m {
        curr_row[0] = i;

        for j in 1..=n {
            let cost = if chars1[i - 1] == chars2[j - 1] { 0 } else { 1 };

            curr_row[j] = cmp::min(
                cmp::min(
                    prev_row[j] + 1,        // deletion
                    curr_row[j - 1] + 1,    // insertion
                ),
                prev_row[j - 1] + cost,     // substitution
            );
        }

        std::mem::swap(&mut prev_row, &mut curr_row);
    }

    prev_row[n]
}

/// Damerau-Levenshtein Distance - Adds transposition operation
/// Time: O(nm), Space: O(nm)
pub fn damerau_levenshtein_distance(str1: &str, str2: &str) -> usize {
    let m = str1.len();
    let n = str2.len();
    let max_dist = m + n;

    let chars1: Vec<char> = str1.chars().collect();
    let chars2: Vec<char> = str2.chars().collect();

    // Use HashMap for sparse matrix
    let mut h: HashMap<(isize, isize), usize> = HashMap::new();

    // Initialize
    h.insert((-1, -1), max_dist);

    for i in 0..=(m as isize) {
        h.insert((i, -1), max_dist);
        h.insert((i, 0), i as usize);
    }

    for j in 0..=(n as isize) {
        h.insert((-1, j), max_dist);
        h.insert((0, j), j as usize);
    }

    for i in 1..=m {
        let mut db = 0;

        for j in 1..=n {
            let k = db;
            let l = if chars1[i - 1] == chars2[j - 1] {
                db = j;
                0
            } else {
                1
            };

            let i_i = i as isize;
            let j_i = j as isize;
            let k_i = k as isize;
            let db_i = db as isize;

            h.insert(
                (i_i, j_i),
                cmp::min(
                    cmp::min(
                        cmp::min(
                            h[&(i_i - 1, j_i)] + 1,          // deletion
                            h[&(i_i, j_i - 1)] + 1,          // insertion
                        ),
                        h[&(i_i - 1, j_i - 1)] + l,          // substitution
                    ),
                    h[&(k_i - 1, db_i - 1)] + (i - k - 1) + 1 + (j - db - 1), // transposition
                ),
            );
        }
    }

    h[&(m as isize, n as isize)]
}

/// Longest Common Subsequence length
/// Time: O(nm), Space: O(nm)
pub fn lcs_length(str1: &str, str2: &str) -> usize {
    let m = str1.len();
    let n = str2.len();

    let chars1: Vec<char> = str1.chars().collect();
    let chars2: Vec<char> = str2.chars().collect();

    let mut dp = vec![vec![0; n + 1]; m + 1];

    for i in 1..=m {
        for j in 1..=n {
            if chars1[i - 1] == chars2[j - 1] {
                dp[i][j] = dp[i - 1][j - 1] + 1;
            } else {
                dp[i][j] = cmp::max(dp[i - 1][j], dp[i][j - 1]);
            }
        }
    }

    dp[m][n]
}

/// Return the actual Longest Common Subsequence
pub fn lcs_string(str1: &str, str2: &str) -> String {
    let m = str1.len();
    let n = str2.len();

    let chars1: Vec<char> = str1.chars().collect();
    let chars2: Vec<char> = str2.chars().collect();

    let mut dp = vec![vec![0; n + 1]; m + 1];

    for i in 1..=m {
        for j in 1..=n {
            if chars1[i - 1] == chars2[j - 1] {
                dp[i][j] = dp[i - 1][j - 1] + 1;
            } else {
                dp[i][j] = cmp::max(dp[i - 1][j], dp[i][j - 1]);
            }
        }
    }

    // Backtrack to find LCS
    let mut lcs = Vec::new();
    let mut i = m;
    let mut j = n;

    while i > 0 && j > 0 {
        if chars1[i - 1] == chars2[j - 1] {
            lcs.push(chars1[i - 1]);
            i -= 1;
            j -= 1;
        } else if dp[i - 1][j] > dp[i][j - 1] {
            i -= 1;
        } else {
            j -= 1;
        }
    }

    lcs.reverse();
    lcs.iter().collect()
}

// ============================================================================
// Similarity Metrics
// ============================================================================

/// Normalized Levenshtein distance (0 to 1)
/// 1.0 = identical, 0.0 = completely different
pub fn normalized_levenshtein(str1: &str, str2: &str) -> f64 {
    if str1.is_empty() && str2.is_empty() {
        return 1.0;
    }

    let distance = levenshtein_distance(str1, str2, false);
    let max_len = cmp::max(str1.len(), str2.len());

    1.0 - (distance as f64 / max_len as f64)
}

/// Similarity ratio based on LCS
pub fn similarity_ratio(str1: &str, str2: &str) -> f64 {
    if str1.is_empty() && str2.is_empty() {
        return 1.0;
    }

    let lcs_len = lcs_length(str1, str2);
    let total_len = str1.len() + str2.len();

    if total_len == 0 {
        return 1.0;
    }

    2.0 * lcs_len as f64 / total_len as f64
}

/// Jaro distance - Good for short strings like names
/// Returns: 0.0 to 1.0 (1.0 = identical)
pub fn jaro_distance(str1: &str, str2: &str) -> f64 {
    if str1 == str2 {
        return 1.0;
    }

    let len1 = str1.len();
    let len2 = str2.len();

    if len1 == 0 || len2 == 0 {
        return 0.0;
    }

    let chars1: Vec<char> = str1.chars().collect();
    let chars2: Vec<char> = str2.chars().collect();

    // Maximum allowed distance
    let mut match_distance = cmp::max(len1, len2) / 2;
    if match_distance > 0 {
        match_distance -= 1;
    }
    if match_distance < 1 {
        match_distance = 1;
    }

    let mut str1_matches = vec![false; len1];
    let mut str2_matches = vec![false; len2];

    let mut matches = 0;

    // Find matches
    for i in 0..len1 {
        let start = if i >= match_distance { i - match_distance } else { 0 };
        let end = cmp::min(i + match_distance + 1, len2);

        for j in start..end {
            if str2_matches[j] || chars1[i] != chars2[j] {
                continue;
            }
            str1_matches[i] = true;
            str2_matches[j] = true;
            matches += 1;
            break;
        }
    }

    if matches == 0 {
        return 0.0;
    }

    // Find transpositions
    let mut transpositions = 0;
    let mut k = 0;

    for i in 0..len1 {
        if !str1_matches[i] {
            continue;
        }
        while !str2_matches[k] {
            k += 1;
        }
        if chars1[i] != chars2[k] {
            transpositions += 1;
        }
        k += 1;
    }

    (matches as f64 / len1 as f64
        + matches as f64 / len2 as f64
        + (matches as f64 - transpositions as f64 / 2.0) / matches as f64)
        / 3.0
}

/// Jaro-Winkler distance - Gives more weight to common prefix
pub fn jaro_winkler_distance(str1: &str, str2: &str, prefix_scale: f64) -> f64 {
    let jaro = jaro_distance(str1, str2);

    // Find common prefix length (max 4)
    let mut prefix = 0;
    let min_len = cmp::min(str1.len(), str2.len());
    let chars1: Vec<char> = str1.chars().collect();
    let chars2: Vec<char> = str2.chars().collect();

    for i in 0..cmp::min(min_len, 4) {
        if chars1[i] == chars2[i] {
            prefix += 1;
        } else {
            break;
        }
    }

    jaro + prefix as f64 * prefix_scale * (1.0 - jaro)
}

// ============================================================================
// Helper Functions
// ============================================================================

fn visualize_dp_table(str1: &str, str2: &str, dp: &[Vec<usize>]) {
    println!("\nDP Table:");
    println!("{}", "=".repeat(70));

    // Header
    print!("       ");
    for c in str2.chars() {
        print!("{:>4}", c);
    }
    println!();

    // Rows
    let chars1: Vec<char> = str1.chars().collect();
    for (i, row) in dp.iter().enumerate() {
        if i == 0 {
            print!("  ");
        } else {
            print!("{:>2}", chars1[i - 1]);
        }

        for val in row {
            print!("{:>4}", val);
        }
        println!();
    }
}

// ============================================================================
// Main Function - Examples
// ============================================================================

fn main() {
    println!("{}", "=".repeat(70));
    println!("EDIT DISTANCE ALGORITHMS - RUST");
    println!("{}", "=".repeat(70));

    // Example 1: Hamming distance
    println!("\n{}", "=".repeat(70));
    println!("EXAMPLE 1: Hamming Distance");
    println!("{}", "=".repeat(70));

    let str1 = "karolin";
    let str2 = "kathrin";

    match hamming_distance(str1, str2) {
        Ok(dist) => {
            println!("String 1: '{}'", str1);
            println!("String 2: '{}'", str2);
            println!("Hamming distance: {}", dist);
        }
        Err(e) => println!("Error: {}", e),
    }

    // Example 2: Levenshtein distance
    println!("\n{}", "=".repeat(70));
    println!("EXAMPLE 2: Levenshtein Distance");
    println!("{}", "=".repeat(70));

    let str3 = "kitten";
    let str4 = "sitting";
    let dist2 = levenshtein_distance(str3, str4, true);

    println!("\nString 1: '{}'", str3);
    println!("String 2: '{}'", str4);
    println!("Levenshtein distance: {}", dist2);

    // Example 3: Damerau-Levenshtein
    println!("\n{}", "=".repeat(70));
    println!("EXAMPLE 3: Damerau-Levenshtein Distance");
    println!("{}", "=".repeat(70));

    let test_cases = vec![
        ("kitten", "sitting"),
        ("teh", "the"),
        ("abcd", "acbd"),
    ];

    for (s1, s2) in test_cases {
        let lev_dist = levenshtein_distance(s1, s2, false);
        let dam_lev_dist = damerau_levenshtein_distance(s1, s2);

        println!("\nString 1: '{}'", s1);
        println!("String 2: '{}'", s2);
        println!("  Levenshtein:         {}", lev_dist);
        println!("  Damerau-Levenshtein: {}", dam_lev_dist);
    }

    // Example 4: LCS
    println!("\n{}", "=".repeat(70));
    println!("EXAMPLE 4: Longest Common Subsequence");
    println!("{}", "=".repeat(70));

    let str5 = "ABCDGH";
    let str6 = "AEDFHR";
    let lcs = lcs_string(str5, str6);

    println!("String 1: '{}'", str5);
    println!("String 2: '{}'", str6);
    println!("LCS: '{}' (length: {})", lcs, lcs.len());

    // Example 5: Similarity metrics
    println!("\n{}", "=".repeat(70));
    println!("EXAMPLE 5: Similarity Metrics");
    println!("{}", "=".repeat(70));

    let test_pairs = vec![
        ("kitten", "sitting"),
        ("saturday", "sunday"),
        ("martha", "marhta"),
        ("dixon", "dicksonx"),
    ];

    for (s1, s2) in test_pairs {
        let norm_lev = normalized_levenshtein(s1, s2);
        let sim_ratio = similarity_ratio(s1, s2);
        let jaro = jaro_distance(s1, s2);
        let jaro_winkler = jaro_winkler_distance(s1, s2, 0.1);

        println!("\n'{}' vs '{}':", s1, s2);
        println!("  Normalized Levenshtein: {:.4}", norm_lev);
        println!("  Similarity Ratio:       {:.4}", sim_ratio);
        println!("  Jaro:                   {:.4}", jaro);
        println!("  Jaro-Winkler:           {:.4}", jaro_winkler);
    }

    println!("\n{}", "=".repeat(70));
}
